tests/schema/test_arguments.py

```diff
@@ -1,10 +1,10 @@
 from textwrap import dedent
-from typing import Optional
+from typing import Optional, Union
 
 from typing_extensions import Annotated
 
 import strawberry
-from strawberry.arguments import UNSET, is_unset
+from strawberry.arguments import UNSET, _Unset, is_unset
```

Comment:
Not sure I like exporting a ~private symbol like this. @patrick91 should we just rename it `Unset`?

Reply:
I'd say yes. But I wonder if we could find a way to make `UNSET`	work as a type, this:

`x: Unset | SomethingElse = UNSET`

feels a bit strange, what do you think @BryceBeagle @sarahhenkens ?

Reply:
Having `UNSET` being set to `any` is the root cause of failures in type checking to occur. Imagine the following scenario:

```python
@strawberry.field
async def jobs(
    self,
    status: Union[_Unset, Optional[Status]] = UNSET,
    first: Optional[int] = UNSET,
    after: Optional[str] = UNSET,
    last: Optional[int] = UNSET,
    before: Optional[str] = UNSET,
) -> Connection[types.Job]:
    result = await Container.job_service().get_paginated(
        status=status, first=first, after=after, last=last, before=before
    )
 ```

From a type checking perspective `UNSET` is technically not valid value for `first: Optional[int] = UNSET,` because its an instance of `_Unset`, but because `UNSET` has its type overwritten to `any`, type checkers will just ignore it.

This gives code a false sense of security, this is why my code example for `status` has an explicit type for it to ensure downstream code is aware of it.

If `Unset` has its correct type set to `_Unset`, type checkers will prevent you from doing  `first: Optional[int] = UNSET,` which is in my opinion a much better default behavior than having `UNSET` be an `any`.

I wanted to keep backward compatibility since other projects might assume it to be any so I just made changes to the internals to support this special type without making changes to the public API and its types.

Reply:
> I'd say yes. But I wonder if we could find a way to make UNSET work as a type, this:
> x: Unset | SomethingElse = UNSET
> feels a bit strange, what do you think @BryceBeagle @sarahhenkens ?

At the core, `x: Optional[T] = UNSET` is essential an 3-way value, "null, not-set, T". In my original ticket (#872) I recommended a new type to represent this tri-state:

```python
T = TypeVar("T")
UnsetOrOptional = Union[_Unset, Optional[T]]
```

This would make the interface much cleaner:
```python
status: UnsetOrOptional[Status]] = UNSET,
```

Reply:
I think I was suggesting to be able to do this:

```
x: UNSET | SomethingElse = UNSET
```

treating UNSET almost like a literal, so wouldn't need to expose `_Unset` ^^

But I'm not sure this is easy to do at the moment!

Reply:
Having a clearer API sound great! 

My changes are just in the core logic to not have it break on its own `_Unset` though.

Thoughts on getting this merged to unblock any projects that want the strict typing while we work in the future on a better general `UNSET` behavior? 

I wanted to keep this PR's scope minimal by just fixing the annotation parser bug.

Reply:
sounds sensible! I'll open an issue for the (potential) improvement :)